caproto/_utils.py

```diff
@@ -602,3 +626,29 @@ def __call__(self):
 
                 self.value = value
                 return value
+
+
+def socket_bytes_available(sock, *, default_buffer_size=4096,
+                           available_buffer=None):
+    '''Return bytes available to receive on socket
+
+    Parameters
+    ----------
+    sock : socket.socket
+    default_buffer_size : int, optional
+        Default recv buffer size, should the platform not support the call or
+        the call fails for unknown reasons
+    available_buffer : array.array, optional
+        Array used for call to fcntl; can be specified to avoid reallocating
+        many times
+    '''
+    if available_buffer is None:
+        available_buffer = array.array('i', [0])
+
+    ok = fcntl.ioctl(sock, termios.FIONREAD, available_buffer) >= 0
+    return (max((available_buffer[0], default_buffer_size))
+            if ok else default_buffer_size)
+
+
+if sys.platform == 'win32' or fcntl is None:
```

Comment:
Defining one version of `socket_bytes_available` _or_ the other is probably better than defining the POSIX one and then, maybe, shadowing it. As written, it's impossible to know what (if anything `_windows_compat` will shadow without cross-referencing.

_Not a criticism of this expedient approach, just a bookmark for one of us to consider cleaning this up._

Reply:
I don't disagree it could be written a lot more clearly, but I also somewhat like shoving all the smelly windows support nonsense in its own corner.